js/ui/workspace.js

```diff
@@ -1556,6 +1556,10 @@ const Workspace = new Lang.Class({
 
     _windowRemoved : function(metaWorkspace, metaWin) {
         this._doRemoveWindow(metaWin);
+
+        // Make sure to show the icon grid if we removed the last window
+        if (!Main.workspaceMonitor.hasActiveWindows)
+            Main.overview.showApps();
```

Comment:
I'm not sure what was going on but all my experiments led to the same point where I could not get the event at the right time unless I did something explicitly like this. But I agree something funny is going on here, will try to get to it later, while doing general cleanups and stabilization.